Implement new checkbox structure for managing provider level permissions

steventux opened the pull request:
## Context

We've iterated the design for managing provider relationship permissions so that we group inputs by permission rather than provider.

<!-- Why are you making this change? What might surprise someone about it? -->

## Changes proposed in this pull request

Revives some of the provider relationship permissions UI temporarily removed in https://github.com/DFE-Digital/apply-for-teacher-training/pull/2520 but using the simpler single permissions model.

Reorganises the edit permissions form so that fields are grouped by permission.
<!-- If there are UI changes, please include Before and After screenshots. -->

![image](https://user-images.githubusercontent.com/93511/88079923-460c1d80-cb76-11ea-89f4-b54bcf8731ea.png)

#### Validating that at least one organisation has permission

![image](https://user-images.githubusercontent.com/93511/88188194-6815a680-cc2f-11ea-9651-13acd6ae8c05.png)


## Guidance to review

The bulk of the complexity in this PR is how we validate seemingly unrelated attributes on the `ProviderRelationshipPermissions` model as pairs of permissions. This makes for a slightly awkward params structure like:

```ruby
[
  { make_decisions: { training_provider_can_make_decisions: 'true', ratifying_provider_can_make_decisions: 'true' },
  { view_safeguarding_information: { ... },
}]
```
This allows us to highlight the correct fieldset in the form when no permissions have been selected.
The last few commits deal with validation so it might make sense to review as separate commits.

<!-- How could someone else check this work? Which parts do you want more feedback on? -->

## Link to Trello card

https://trello.com/c/g4jGI6XK/2457-implement-new-checkbox-structure-for-managing-provider-level-permissions

<!-- http://trello.com/123-example-card -->

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training#azure-hosting-devops-pipeline)

duncanjbrown commented on app/controllers/provider_interface/provider_relationship_permissions_controller.rb:
```diff
@@ -0,0 +1,67 @@
+    end
+
+    def provider_relationship_params
+      params.permit(:ratifying_provider_id, :training_provider_id).to_h
```
Given that all these relationships will exist in the database anyway, would it be simpler to use `:relationship_id`?

steventux replied:
Yes I think so, the only actions that flow into this controller logic will come from the organisations pages which already lookup permissions records or from the setup redirect which have to retrieve these records anyway.

steventux replied:
@duncanjbrown https://github.com/DFE-Digital/apply-for-teacher-training/pull/2547/commits/45c09755185c84227644894e4aa87ab68a624558 should cover routing to these actions via the permissions ID.

adamsilver commented on spec/system/provider_interface/manage_provider_relationship_permissions_spec.rb:
```diff
@@ -0,0 +1,158 @@
+  end
+
+  def then_i_can_see_a_validation_error_telling_me_to_assign_an_org_to_a_permission
+    expect(page).to have_content('At least one organisation must have permission to view safeguarding information')
```
```suggestion
    expect(page).to have_content('Select which organisations can view safeguarding information')
```

adamsilver commented on app/models/provider_relationship_permissions.rb:
```diff
@@ -11,4 +22,14 @@ def training_provider_can_view_applications_only?
+  def at_least_one_active_permission_in_pair
+    PERMISSIONS.each do |permission|
+      if !send("training_provider_can_#{permission}") && !send("ratifying_provider_can_#{permission}")
+        errors.add(permission, "At least one organisation must have permission to #{permission.to_s.humanize.downcase}")
```
```suggestion
        errors.add(permission, "Select which organisations can #{permission.to_s.humanize.downcase}")
```

steventux commented:
@duncanjbrown I've incorporated your suggestion about reworking params and rebased into https://github.com/DFE-Digital/apply-for-teacher-training/pull/2547/commits/78d6170469ff1773e22fb308fd8b17a308fdda3d

duncanjbrown commented on app/models/provider_relationship_permissions.rb:
```diff
@@ -2,7 +2,18 @@ class ProviderRelationshipPermissions < ApplicationRecord
+
+  validate :at_least_one_active_permission_in_pair, if: -> { setup_at.present? }
+
+  def self.permissions_fields
```
Do we use this anywhere?

steventux replied:
@duncanjbrown Regarding validation - _I think_ the heroku provider relationships all have `setup_at: nil`, the initial state of a provider-to-provider relationship, this means the validation won't add any errors. Validation works for me locally.
We should never have this situation once the redirect to the permissions setup screens are back in place as this process sets that date.
I've removed the `permissions_fields` method.
